### ISSUE TITLE
Unable to generate proxy for class that implements multiple interfaces.

### ISSUE THREAD

**Issue by mminella** [spring-projects-experimental/spring-native#459]
In Spring Batch, there are two custom scopes `StepScope` and `JobScope`. In each of these scenarios, the bean's definition is proxied so that the underlying bean can be created later during execution. This functionality allows for things like the ability to inject job parameters into Spring Batch components.

However, the proxies being created when using a native image are based on the return type of a method defined with the `@Bean` annotation. If I have a class that implements multiple interfaces that must be exposed via the poxy, there is no way currently to generate that.

An example of the issue can be found here: https://github.com/mminella/spring-graalvm-native/tree/batch_io. In that example, in the `BatchConfiguration` class, there is a bean defined as:

```java 
@StepScope
@Bean
public ItemReaderListener reader() {
   ...
}
```

`ItemReaderListener` is a class that implements the `ItemReader`, `ItemStream`, and the `StepExecutionListener` interfaces. Spring Batch provides a composite interface for `ItemReader` and `ItemStream` so the method can be refactored to return `ItemStreamReader`. However, the proxy created based on that definition cannot be cast to a `StepExecutionListener`.

**Comment by sdeleuze** [comment]
Any thoughts on that one @aclement?

**Comment by aclement** [comment]
I wrestled with this one (with Michael) last week. My initial reaction was to make the proxy creation we have in spring a little smarter, but I haven't dug into it.  Right now we have that patch which says "if not native-image do the right thing, otherwise just give them a jdkproxy", I wondered about that native-image clause, making it do a little more analysis (certainly produce more diagnostics because when a bad situation arises it causes chaos downstream trying to debug).

**Comment by sdeleuze** [comment]
So we should have something more clever that tries to create a JDK proxy configuration for the various interfaces implemented by the related class?

**Comment by mminella** [comment]
The creating a JDK proxy configuration for the various interfaces is what I thought I was attempting to do in my `BatchScopeComponentProcessor` here: https://github.com/mminella/spring-graalvm-native/blob/batch_io/spring-native-configuration/src/main/java/org/springframework/batch/core/configuration/annotation/BatchScopeComponentProcessor.java but to no avail.

**Comment by aclement** [comment]
> So we should have something more clever that tries to create a JDK proxy configuration for the various interfaces implemented by the related class?

Something like that. But it is mixed up with when to fail fast - if the 'original spring' really wanted to create an objenesis proxy then maybe that's when we should fail fast. But how we'd recognize additionally "oh its spring batch, it just wants a proxy for that set of interfaces rather than a further proxy of the class, so let's just the interfaces of this class and return a proxy for those" I'm not sure - I haven't put enough brain power in yet but there could be a way through.  Re Michaels comment - I think creating it is one thing but that isn't the proxy we go looking for with the current native-image block in Spring, so it doesn't work (I *think*)

**Comment by aclement** [comment]
@sdeleuze I got something working vaguely along the lines I was thinking of. What do you think about this?
```
	public AopProxy createAopProxy(AdvisedSupport config) throws AopConfigException {
		// Check if this would have been an objenesis proxy rather than a JDK proxy
		// These conditions are the same as the paths taken in the real createAopProxy
		boolean a = config.isOptimize() || config.isProxyTargetClass() || hasNoUserSuppliedProxyInterfaces(config);
		Class<?> tc = config.getTargetClass();
		boolean b = tc!=null && (tc.isInterface() || Proxy.isProxyClass(tc));
		if (a && !b) { // The original impl would have been an objenesis class proxy
			// Add the interfaces of the target class to the list which will be proxied:
			if (tc != null && !tc.isInterface()) {
				List<Class<?>> intfaces = new ArrayList<>();
				Helper.collect(tc,intfaces);
				for (Class<?> intface: intfaces) {
					config.addInterface(intface);
				}
			}
		}
		return new JdkDynamicAopProxy(config);
	}
```
Basically it says "if we would have returned an objenesis proxy for a class target then lets at least try to return a more representative JDK proxy by adding the interfaces in the hierarchy of the target class to the proxy configuration".  It may not be fully polished, but it does work for the batch case, and it seems better than what we do now.

**Comment by sdeleuze** [comment]
Yes, I think that's better and useful for Spring Framework 6 preparation cc @jhoeller.

**Comment by sdeleuze** [comment]
@aclement Does this upcoming change means that we should close #221 (Throw an exception if `@Transactional` is not in an interface) without implementing it?

**Comment by sdeleuze** [comment]
Should we move this issue to 0.9.1? See [this related comment](https://github.com/spring-projects-experimental/spring-native/issues/203#issuecomment-778995541).

**Comment by sdeleuze** [comment]
@aclement Any thoughts?

**Comment by aclement** [comment]
The ability to create proxies that subtype classes was something done a while ago. This gave us the ability to then write hints that could express these needs (like the one you reference).  What AOT has come along and done has been to be smarter about inferring these hints and sometimes push work to build time such that the hints are not required for runtime.  So on the first point:

> The batch-io sample seems to be working with the latest main (as of https://github.com/spring-projects-experimental/spring-native/commit/afc0d75fd2b6cec381fd88552570a9956f159850) without the proxy hint.

Yep, could be, could be that we just haven't gone back to review all the samples as the AOT system has been enhanced to delete hints that are now inferred automatically.

**Comment by sdeleuze** [comment]
@aclement From you POV should we close this issue?

**Comment by aclement** [comment]
Yes, it could be closed - for the subject of the issue we do have the mechanism now. The only possible bit of work could be to delete the manual hints that @benas eludes to in the samples if they are no longer required.

**Comment by sdeleuze** [comment]
The sample is fixed in 0.12.0 but the original issue was probably fixed in 0.11.x.
